tests/test_gang.F90

```diff
@@ -0,0 +1,798 @@
+! In applying this licence, ECMWF does not waive the privileges and immunities
+! granted to it by virtue of its status as an intergovernmental organisation
+! nor does it submit to any jurisdiction.
+
+PROGRAM TEST_GANG
+
+USE FIELD_MODULE
+USE FIELD_FACTORY_MODULE
+USE FIELD_ACCESS_MODULE
+USE FIELD_ABORT_MODULE
+USE PARKIND1, ONLY : JPRD, JPIM
+
+IMPLICIT NONE
+
+CLASS(FIELD_4RD), POINTER :: YLF4
+TYPE(FIELD_3RD_PTR), ALLOCATABLE :: YLF3L (:)
+
+INTEGER (KIND=JPIM), PARAMETER :: NPROMA = 32, NFLEVG = 15, NDIM = 3, NGPBLKS = 10
+LOGICAL, PARAMETER :: LLPERSISTENT = .TRUE.
+
+INTEGER (KIND=JPIM), PARAMETER :: ILB (4) = [1,0,1,1], IUB (4) = [NPROMA, NFLEVG, NDIM, NGPBLKS]
+REAL (KIND=JPRD), ALLOCATABLE :: ZDATA4 (:,:,:,:)
+
+INTEGER (KIND=JPIM) :: JDIM
+INTEGER (KIND=JPIM) :: ISTEP
+LOGICAL, PARAMETER :: LLVERBOSE = .TRUE.
+
+INTEGER :: IERROR = 0
+
+REAL (KIND=JPRD) :: ZCOEF (NDIM)
+!$acc declare create (ZCOEF)
+
+ALLOCATE (ZDATA4 (ILB (1):IUB (1), ILB (2):IUB (2), ILB (3):IUB (3), ILB (4):IUB (4)))
+
+DO JDIM = 1, NDIM
+  ZCOEF (JDIM) = REAL (JDIM, JPRD)
+ENDDO
+
+YLF4 => NULL ()
+
+DO ISTEP = 1, 2000
+  CALL STEP
+ENDDO
+
+IF (ASSOCIATED (YLF4)) THEN
+  CALL DEL_YLF4
+ENDIF
+
+CONTAINS
+
+SUBROUTINE TEST_GANG_ABORT (CDMESS)
+
+CHARACTER (LEN=*), INTENT (IN) :: CDMESS
+
+IERROR = IERROR + 1
+
+END SUBROUTINE
+
+INTEGER (KIND=JPIM) FUNCTION IRAND (K1, K2)
+
+INTEGER (KIND=JPIM) :: K1, K2
+
+INTEGER*8, SAVE :: X = 2713
+
+X = MODULO (16807_8 * X, 2147483647_8)
+
+IRAND = K1 + MODULO (X, INT (K2-K1+1, 8))
+
+END FUNCTION
+
+SUBROUTINE SORTR8 (KDIM, KORD, PVAL)
+
+INTEGER (KIND=JPIM) :: KDIM
+INTEGER (KIND=JPIM) :: KORD (KDIM)
+REAL (KIND=JPRD)    :: PVAL (KDIM)
+
+INTEGER (KIND=JPIM) :: I, J, ITMP
+
+DO I = 1, KDIM
+  DO J = I+1, KDIM
+    IF (PVAL (KORD (I)) > PVAL (KORD (J))) THEN
+      ITMP = KORD (I)
+      KORD (I) = KORD (J)
+      KORD (J) = ITMP
+    ENDIF
+  ENDDO
+ENDDO
+
+END SUBROUTINE
+
+SUBROUTINE STEP
+
+INTEGER (KIND=JPIM) :: IFLD (NDIM)
+INTEGER (KIND=JPIM) :: ICASE, ISIZE
+INTEGER (KIND=JPIM) :: JFLD, JDIM, ITMP
+
+REAL (KIND=JPRD) :: ZVAL (NDIM)
+
+ISIZE = IRAND (1, NDIM)
+
+DO JFLD = 1, NDIM
+  IFLD (JFLD) = JFLD
+  ZVAL (JFLD) = REAL (IRAND (0, 100000), JPRD) / REAL (100000, JPRD)
+ENDDO
+
+
+CALL SORTR8 (NDIM, IFLD, ZVAL)
+
+IF (.NOT. ASSOCIATED (YLF4)) THEN
+
+  SELECT CASE (IRAND (1, 2))
+    CASE (1)
+      CALL NEW_OWNER_YLF4
+    CASE (2)
+      CALL NEW_WRAPPER_YLF4
+    CASE DEFAULT
+      CALL FIELD_ABORT ('UNEXPECTED CASE')
+  END SELECT
+  
+  SELECT CASE (IRAND (1, 6))
+    CASE (1)
+      CALL SET_VIEW_YLF4
+    CASE (2)
+      CALL SET_DEVICE_YLF4
+    CASE (3)
+      CALL SET_HOST_YLF4
+    CASE (4)
+      CALL SET_VIEW_YLF3 ([(JFLD, JFLD = 1, NDIM)])
+    CASE (5)
+      CALL SET_DEVICE_YLF3 ([(JFLD, JFLD = 1, NDIM)])
+    CASE (6)
+      CALL SET_HOST_YLF3 ([(JFLD, JFLD = 1, NDIM)])
+    CASE DEFAULT
+      CALL FIELD_ABORT ('UNEXPECTED CASE')
+  END SELECT
+  
+ELSE
+
+  SELECT CASE (IRAND (1, 134))
```

Comment:
I'm not sure I understand the use of `IRAND` here. Are we selecting the case to test based on an RNG, or am I missing something here?

Reply:
Yes, we generate random (but reproducible) sequences of operations on the 4D fields and its 3D children. Note it is not feasible to explore all the combinations.

Reply:
For the record, I don't think RNG-based testing is a good idea, and I still don't fully understand the need for this here. However, fixing this probably requires a rethink of the test base implementation, which is beyond this PR.